br/pkg/lightning/duplicate/worker.go

```diff
@@ -0,0 +1,203 @@
+				return nil
+			}
+			if err := w.runTask(ctx, t); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (w *worker) runTask(ctx context.Context, t task) (retErr error) {
+	logTask := w.logger.With(
+		zap.Stringer("startKey", t.startKey),
+		zap.Stringer("initialEndKey", t.endKey),
+	).Begin(zap.InfoLevel, "[dup-detector] runTask")
+
+	var processedKeys int64
+
+	defer func() {
+		w.taskWg.Done()
+		err := w.handler.Close()
+		if retErr == nil {
+			retErr = err
+		}
+		logTask.End(zap.ErrorLevel, retErr,
+			zap.Stringer("endKey", t.endKey),
+			zap.Int64("processedKeys", processedKeys),
+		)
+	}()
+
+	it, err := w.sorter.NewIterator(ctx)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = it.Close()
+	}()
+
+	// Periodically check context cancellation or task splitting during iteration.
+	const checkInterval = 1000
+	numIterations := 0
+
+	inDup := false
+	var prevKey, curKey internalKey
+	for it.Seek(encodeInternalKey(nil, t.startKey)); it.Valid(); it.Next() {
+		if err := decodeInternalKey(it.UnsafeKey(), &curKey); err != nil {
+			return err
+		}
+		if compareInternalKey(curKey, t.endKey) >= 0 {
+			break
+		}
+		processedKeys++
+
+		if bytes.Equal(curKey.key, prevKey.key) {
+			if inDup {
+				if err := w.handler.Append(curKey.keyID); err != nil {
+					return err
+				}
+			} else {
+				if err := w.handler.Begin(curKey.key); err != nil {
+					return err
+				}
+				if err := w.handler.Append(prevKey.keyID); err != nil {
+					return err
+				}
+				if err := w.handler.Append(curKey.keyID); err != nil {
+					return err
+				}
+				inDup = true
+				w.numDups.Add(1)
+			}
+		} else if inDup {
+			if err := w.handler.End(); err != nil {
+				return err
+			}
+			inDup = false
+		}
+
+		numIterations++
+		// Try to split the task and send the new task to the task channel
+		// so that other idle workers can process it.
+		if numIterations%checkInterval == 0 {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+			if len(w.taskCh) == 0 {
+				userSplitKey := genSplitKey(curKey.key, t.endKey.key)
+				if bytes.Compare(userSplitKey, curKey.key) > 0 {
+					splitKey := internalKey{key: userSplitKey}
+					w.taskWg.Add(1)
+					select {
+					case w.taskCh <- task{startKey: splitKey, endKey: t.endKey}:
+						t.endKey = splitKey
+					default:
+						w.taskWg.Done()
+					}
+				}
+			}
+		}
+		prevKey, curKey = curKey, prevKey
+	}
+	if err := it.Error(); err != nil {
+		return err
+	}
+	if inDup {
+		if err := w.handler.End(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func genSplitKey(startKey, endKey []byte) []byte {
+	if bytes.Equal(startKey, endKey) {
+		return slices.Clone(startKey)
+	}
+
+	l := commonPrefixLen(startKey, endKey)
+	splitKey := slices.Clone(startKey[:l])
+
+	if l == len(startKey) {
+		// The start key is a prefix of the end key.
+		splitKey = append(splitKey, endKey[l]/2)
+		return splitKey
+	}
+
+	c1, c2 := startKey[l], endKey[l]
```

Comment:
will endKey equal to empty slice? so we will meet index out of range

Reply:
It's guaranteed that `endKey` is greater than `startKey` here.